### PR TITLE
Modify `pjson` format

### LINKED ISSUE #1581: Modify `pjson` format
Currently, `pjson` stores the value's type together with the value itself:

```json
{
    "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
    "doc": {
        "$k": ["s"],
        "s": "foo"
    },
    "array": [1, 2],
    "d": {"$f": 42.13},
    "s": "bar",
    "bin": {"$b": "YmF6", "s": 0},
    "oid": {"$o": "010101010101010101010101"},
    "bool": true,
    "t": {"$d": -62167219200000},
    "null": null,
    "regex": {"$r": "hoffman", "o": "i"},
    "i32": 42,
    "ts": {"$t": "1"},
    "i64": {"$l": "9223372036854775807"}
}
```

That limits our ability to pushdown queries and index that data effectively. For example, in queries, MongoDB often does not differentiate between numeric types, but that format does.

Let's extract the whole document's schema into its own field and store values as simpler JSON:

```json
{
    "$s": {
      "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
      "p": { 
        "doc": {
            "t": "object",
            "$s": {
                "$k": ["s"],
                "s": {"t": "string"}
            }
        },
        "array": {
            "t": "array",
            "i": ["int", "int"]
        },
        "d": {"t": "double"},
        "s": {"t": "string"},
        "bin": {"t": "binData", "s": 0},
        "oid": {"t": "objectId"},
        "bool": {"t": "bool"},
        "t": {"t": "date"},
        "null": {"t": "null"},
        "regexp": {"t": "regex","o": "i"},
        "i32": {"t": "int"},
        "ts": {"t": "timestamp"},
        "i64": {"t": "long"}
      }
    },
    "doc": {
        "s": "foo"
    },
    "array": [1, 2],
    "d": 42.13,
    "s": "bar",
    "bin": "YmF6",
    "oid": "010101010101010101010101",
    "bool": true,
    "t": -62167219200000,
    "null": null,
    "regex": "hoffman",
    "i32": 42,
    "ts": 1,
    "i64": 9223372036854775807
}
```

Notes:

- [x] subdocuments don't have their own schema; it is a part of the main document schema;
- [x] array's schema is decided
- [x] value types currently match MongoDB type aliases, but that might change in the future; we should not try to reuse those constants;
- [x] scalar values are stored as JSON scalars; their options are a part of the schema;
- [x] int64 values should be stored as numbers; PostgreSQL could handle them.
- [x] 0nly `pjson` and `pgdb` (for query pushdown) packages should be updated. If something else should be changed, that's a bug that should be either fixed or a ticket should be created.
- [x] `fjson` package should not be updated. It is not related to `pjson` anymore.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #1581.

⚠️ The changes in this PR break backward compatibility. Please dump the existing databases before rolling out the changes and restore them afterwards.

## Readiness checklist

* [x] I added tests for new functionality or bugfixes.
* [x] I ran `task all`, and it passed.
* [x] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [inline internal/handlers/pg/pjson/pjson.go:303]: I don't really like that we have both `Marshal` and `MarshalElem` here, but both are needed.

`Marshal` is used to make a document for insert and update commands.
`MarshalElem` is used to make a value for `WHERE` clause. 

If you have an idea for a better design, please propose it.

**Comment by AlekSi** [comment]
Please update CHANGELOG.md

**Comment by rumyantseva** [comment]
@Fashander in this PR we make changes that break backward compatibility, so I need your review of changelog to ensure that it's clear and user-friendly.

**Review by Fashander — APPROVED** [review]
LGTM

**Review by chilagrow — COMMENTED** [review]
Comments so far.

File: internal/handlers/pg/pjson/array.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/array.go:56]: Well, in principle we don't test lazyerrors.

File: internal/handlers/pg/pjson/datetime.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/datetime.go:39]: Yes. And json marshalling/unmarshalling of values is still in use (just like for int64 itself).

File: internal/handlers/pg/pjson/document.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/document.go:122]: Nice catch!

File: internal/handlers/pg/pjson/pjson_test.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson_test.go:257]: Do you have a particular suggest for naming? I was trying to keep it more or less consistent with `tjson`, see the func `unmarshalJSON` in `tjson_test.go`

File: internal/handlers/pg/pjson/schema.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema.go]: @chilagrow in the DB, we don't store them in plain text, they are stored as `jsonb` values anyway, so with that whitespaces don't matter.

File: internal/handlers/pg/pjson/array.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/array.go:49]: The idea is to have different messages to help understand the root cause of the problem. If such problem happens, the value hasn't been marshaled correctly (and this is why we can't unmarshal it), so this is not a normal/correct error, it's a sign that things go wrong. My idea here is to help simplify debugging when things go wrong.

I believe we don't need a test for it because normally we don't test lazyerrors.

File: internal/handlers/pg/pjson/array.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/array.go:56]: If needed, I can write unit tests for such cases, but I'm not sure if it makes sense.

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go:36]: Fixed!

File: internal/handlers/pg/pjson/pjson.go
Comment (w84thesun) [inline internal/handlers/pg/pjson/pjson.go:360]: Maybe we can name it `MarshalSingleValue`? It is precisely what it does and what the docs say.

File: internal/handlers/pg/pjson/schema.go
Comment (w84thesun) [inline internal/handlers/pg/pjson/schema.go:126]: Maybe we can call it `marshalSchema` as it converts the document to a slice of bytes?

File: internal/handlers/pg/pjson/schema.go
Comment (w84thesun) [inline internal/handlers/pg/pjson/schema.go:173]: Maybe we can call it `marshalValueSchema`?

Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema.go:173]: `elem` is how schema elem data type called (see in the beginning of this file), this is why I wanted to call it `elem`. However, `elem` itself is not a good name either, so maybe it needs to be renamed too.

File: internal/handlers/pg/pjson/schema.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema.go]: Suggested a new name (and docs)

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: Fixed

Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: Awww, that suggestion gives me naming idea :) 

- `pjson.Marshal` to create a top-level json object that includes the document and the schema
- `pjson.Unmarshal` to unmarshal that top-level json object
- `MarshalSingleValue` to marshal a single value
- `unmarshalSignleValue` to unmarshal single value with the given `elem` 
- `marshalSchemaForDoc` to marshal schema based on the given doc 
- `marshalElemForSingleValue` to marshal elem for the given single value

Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: fixed

File: internal/handlers/pg/pjson/array.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/array.go:49]: After some discussion, I added a test case to cover this. 
The reason for the test case is to check that we don't ignore such errors (which could lead to a wrong interpretation of data).

File: internal/handlers/pg/pjson/array.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/array.go:56]: After some discussion, I added a test case to cover this.
The reason for the test case is to check that we don't ignore such errors (which could lead to a wrong interpretation of data).

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [inline internal/handlers/pg/pjson/pjson.go:75]: That was a bug - `dr.Len()` is zero in this case, as after `ReadAll`, nothing is left there.

File: internal/handlers/pg/pjson/pjson.go
Comment (chilagrow) [inline internal/handlers/pg/pjson/pjson.go:214]: Not exactly this, but we can be specific about schema keys and document keys mismatch :hugs:
```suggestion
		return nil, lazyerrors.Errorf("pjson.Unmarshal: document must have the same number of keys as schema (schema: %d, keys: %d)", len(sch.Keys), len(v))
```

File: internal/handlers/pg/pjson/pjson_test.go
Comment (chilagrow) [reply internal/handlers/pg/pjson/pjson_test.go:257]: Let's keep it as it is. The consistency is always the way to go when there is a doubt :hugs:

**Review by chilagrow — COMMENTED** [review]
The most important comment is about `$k` vs `keys` I think I missed or forgot an earlier discussion. Other comments are minor. Thanks a lot for amazing work on the testing and documentation as well :pray:

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: Nice catch!

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: fixed

File: internal/handlers/pg/pjson/document.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/document.go]: fixed

File: internal/handlers/pg/pjson/binary.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/binary.go]: fixed

Comment (rumyantseva) [reply internal/handlers/pg/pjson/binary.go]: fixed

File: internal/handlers/pg/pjson/pjson.go
Comment (rumyantseva) [reply internal/handlers/pg/pjson/pjson.go]: fixed

**Review by chilagrow — APPROVED** [review]

**Review by w84thesun — APPROVED** [review]
Nice!

**Review by w84thesun — APPROVED** [review]

**Comment by rumyantseva** [comment]
@AlekSi as I modified changelog, this PR needs your explicit approval too (because of https://github.com/FerretDB/FerretDB/blob/main/.github/CODEOWNERS)

**Review by noisersup — APPROVED** [review]
There's a lot of done here! Looks good to me!

**Review by chilagrow — APPROVED** [review]

**Review by Fashander — APPROVED** [review]

**Pull request merged**
